### PR TITLE
Kafka contain deps

### PR DESCRIPTION
**Pull request by GCHQ-11**
This pulls all the specific dependencies on Kafka into a single implementation library, which can be included at Runtime to make itself available.
I have created a Stroom abstraction for producing to Kafka, which uses Java ServiceLoader to find an implementation.
At this stage I believe it's all still in a single classloader, so it will not necessarily support multiple versions of Kafka at the same time, but it should make it easier to swap specific versions in and out without potentially needing compilation

### CODE REVIEW

**Comment by at055612** [comment]
We need to consider if we want to change the arguments to the send method on our producer abstraction.  Currently it has a consumer of an Exception.  The current kafka impl has a callback that gives you a Metadata object.  Maybe we want some abstraction of this but that could be hard to do without making it too close to one impl.  What we do about this depends on whether stroom cares about the metadata coming back or not.

I think the StroomKafkaProducerFactory.getProducer method should also take something like a java.util.Properties object or Map<String,Object> so that we can pass various props to the impl, which also allows for a bit of extensibility.  If we have a mechanism of stripping some prefix off a stroom property then we can effectively pass kafka's own property keys right through. E.g. stroom.kafkaProducer.buffer.memory becomes buffer.memory when passed through.

It would be nice to remove references to StroomLogger in the impl module and for it to just use plain old SLF4j.

**Comment by GCHQ-11** [comment]
As discussed, we weren't sure about making a copy of the metadata return value, since we aren't using it currently in Stroom I am inclined to leave our interface as is, unless we think of some reason to adjust it.

Properties - yep good idea. I will do this.

StroomLogger - I think the 0.10.0.1 provider library is already only tied to SLF4j

I will also change the KafkaProducer back to a singleton, as they are thread safe and expensive. This makes it more equivalent to what we had before I got my mits on it.

File: stroom-connectors/src/main/java/stroom/connectors/ConnectorProperties.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * Extract the String value of a property for a given key.
+     *
+     * @param key The property key
+     * @return The value of the property as a string.
+     */
+    String getProperty(String key);
+
+    /**
+     * Extract the String value of a property for a given key.
```
Comment (at055612) [inline stroom-connectors/src/main/java/stroom/connectors/ConnectorProperties.java]: copy/paste

File: stroom-connectors/src/main/java/stroom/connectors/ConnectorProperties.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ * that expect prefixed properties to be available in StroomPropertyService.
+ */
+public interface ConnectorProperties {
+    /**
+     * Extract the String value of a property for a given key.
+     *
+     * @param key The property key
+     * @return The value of the property as a string.
+     */
+    String getProperty(String key);
```
Comment (at055612) [inline stroom-connectors/src/main/java/stroom/connectors/ConnectorProperties.java]: I would be inclined to use Optional<...> for the return type of the 3 get methods.  Big fan of optional.  Nice and explicit about what you are getting back.

File: stroom-core-server/src/main/java/stroom/connectors/kafka/StroomKafkaProducerFactoryService.java [hunk 3]
```diff
@@ -32,14 +35,14 @@
     private static final String DEFAULT_NAME = "default";
 
     // Old configs may not have named server/version properties, so use this property name
-    private static final String PROP_BOOTSTRAP_SERVERS = "stroom.kafka.bootstrap.servers";
+    private static final String OLD_PROP_BOOTSTRAP_SERVERS = "stroom.kafka.bootstrap.servers";
```
Comment (at055612) [inline stroom-core-server/src/main/java/stroom/connectors/kafka/StroomKafkaProducerFactoryService.java]: kafka was only introduced in v6 which is still unreleased.  I would remove this backwards compatibility

File: stroom-core-server/src/main/java/stroom/connectors/kafka/StroomKafkaProducerFactoryService.java [hunk 4]
```diff
@@ -32,14 +35,14 @@
...
 
     // Default kafka client version for old configs
     private static final String KAFKA_CLIENT_VERSION_DEFAULT = "0.10.0.1";
```
Comment (at055612) [inline stroom-core-server/src/main/java/stroom/connectors/kafka/StroomKafkaProducerFactoryService.java]: see comment above

File: stroom-core-server/src/main/java/stroom/connectors/kafka/StroomKafkaProducerFactoryService.java [hunk 5]
```diff
@@ -32,14 +35,14 @@
...
```
Comment (at055612) [inline stroom-core-server/src/main/java/stroom/connectors/kafka/StroomKafkaProducerFactoryService.java]: The defaults for all our properties live in stroomCorePropertyContext.  Have you set them in there?  Also there is an  entry for the bootstrap servers in stroom.conf.example in the root of the repo.

File: stroom-core-server/src/main/java/stroom/connectors/ExternalLibService.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    private URL toURLSafe(final URI uri) {
+        URL url = null;
+
+        try {
+            url = uri.toURL();
+        } catch (MalformedURLException e) {
+            final String msg = String.format("Could not load in an External Service JAR %s: %s",
+                    uri,
+                    e.getLocalizedMessage());
+            LOGGER.error(msg, e);
```
Comment (at055612) [inline stroom-core-server/src/main/java/stroom/connectors/ExternalLibService.java]: for info, you should be able to do 
LOGGER.error("some text {} and {}", var1, var2, e)
Saves the extra string format
We also have LammbdaLogger for when constructing the message is expensive.

**Pull request closed without merging**